### DISCUSSION TITLE
Push rtmp use CMSampleBuffer

### DISCUSSION THREAD

**Discussion by youtengli** [HaishinKit/HaishinKit.swift#1009] [Q&A] [answered]
Hello everyone,

I'm a newbie in swift.

I follow an example from [DeepAR Sdk](https://github.com/DeepARSDK/videocall-ios-swift/blob/4fe3aed58a7bed5b579445d26e19a803a7b2e211/videocall-ios-swift/ViewController.swift#L264).

This example pushes CMSampleBuffer to [Agora](https://www.agora.io/en/), but I want to push DeepAR's result to my own rtmp server.

How to push rtmp using CMSampleBuffer?

**Comment by onebuckgames** [comment]
Hi. You can create an effect inherited from the VideoEffet and process your buffer in method

override func execute(_ image: CIImage, info: CMSampleBuffer?) -> CIImage

**Comment by youtengli** [comment]
Hi,

Thank you for your reply.

Can you give me more information?

I have create MyEffect.

But I don't know what to do in execute.

```
import HaishinKit
import AVFoundation

class MyEffect: VideoEffect {
	
	override func execute(_ image: CIImage, info: CMSampleBuffer?) -> CIImage{
		
		image
	}
	
}
```

**Comment by onebuckgames** [comment, marked as answer]
You should store and keep updated your CMSampleBuffer from DeepAR as member of MyEffect, or you can render CIImage from CMSampleBuffer and keep it as a buffer in MyEffect, then in execute method you just return it as is.
